### PR TITLE
[FIXED JENKINS-33319] Premtively wake the acceptor thread to let it close cleanly

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-33319](https://issues.jenkins-ci.org/browse/JENKINS-33319)
- Also adds a Ping Agent protocol that could be used by nagios monitoring, etc. to verify that the slave agent listener is alive
- We use the ping agent protocol to ensure that the acceptor thread wakes up, loops and sees that the shutdown is started
  that prevents the socket close exception from being thrown

```
Mar 04, 2016 10:02:07 AM hudson.TcpSlaveAgentListener shutdown
WARNING: Failed to close down TCP port
java.io.IOException: No such file or directory
    at sun.nio.ch.NativeThread.signal(Native Method)
    at sun.nio.ch.ServerSocketChannelImpl.implCloseSelectableChannel(ServerSocketChannelImpl.java:283)
    at java.nio.channels.spi.AbstractSelectableChannel.implCloseChannel(AbstractSelectableChannel.java:234)
    at java.nio.channels.spi.AbstractInterruptibleChannel.close(AbstractInterruptibleChannel.java:115)
    at hudson.TcpSlaveAgentListener.shutdown(TcpSlaveAgentListener.java:119)
    at jenkins.model.Jenkins.cleanUp(Jenkins.java:2789)
...
```

@reviewbybees

### CODE REVIEW

**Comment by amuniz** [comment]
:bee: (assuming tests pass)

File: core/src/main/java/hudson/TcpSlaveAgentListener.java [hunk 1]
```diff
@@ -116,6 +123,16 @@ public void run() {
     public void shutdown() {
         shuttingDown = true;
         try {
+            SocketAddress localAddress = serverSocket.getLocalAddress();
```
Comment (jtnord) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: Hmm the  ServerSocketChannel should really be in [non blocking mode](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/spi/AbstractSelectableChannel.html#configureBlocking(boolean)) and should check for a null return and wait appropriately (and then all you ned here is a notify).

File: core/src/main/java/hudson/TcpSlaveAgentListener.java [hunk 2]
```diff
@@ -184,6 +201,78 @@ private void error(PrintWriter out, String msg) throws IOException {
...
+     * @since
+     */
+    @Extension
+    public static class PingAgentProtocol extends AgentProtocol {
+
+        private final byte[] ping;
+
+        public PingAgentProtocol() {
+            try {
+                ping = "Ping\n".getBytes("UTF-8");
```
Comment (jtnord) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: I would like to introduce you to `String.getBytes(Charset)` and `StandardCharsets.html.UTF_8` - they will be your new best friends!

"Ping\n".getBytes([StandardCharsets.UTF_8](http://docs.oracle.com/javase/7/docs/api/java/nio/charset/StandardCharsets.html#UTF_8))

**Comment by jtnord** [comment]
I think this is actually the wrong way to do it here.  

I believe the better fix is to use non blocking IO, leaving the ping command to a different PR.

**Comment by stephenc** [comment]
@jtnord The issue here is that the accept logic is in blocking mode because the NIO support was added after the `AgentProtocol` abstraction was introduced.

As a result `AgentProtocol` implementations are to assume that the socket they are asked to handle is in blocking mode... and then they switch it over to non-blocking mode (where the backing socket can handle that)...

The whole thing is driving me nuts, but this seems to be the only way without breaking backwards compatibility contract of `AgentProtocol`

File: core/src/main/java/hudson/TcpSlaveAgentListener.java (hunk 2 above)
Comment (stephenc) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: I'm hoping to backport this as far as 1.609... which is Java 6

File: core/src/main/java/hudson/TcpSlaveAgentListener.java (hunk 1 above)
Comment (stephenc) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: Well if I were writing this from scratch, yep... but that's not how `AgentProtocol` works :cry:

**Comment by jtnord** [comment]
:bee:

**Comment by stephenc** [comment]
@reviewbybees done

**Comment by stephenc** [comment]
@jenkinsci/code-reviewers

File: core/src/main/java/hudson/TcpSlaveAgentListener.java [hunk 3]
```diff
@@ -184,6 +201,78 @@ private void error(PrintWriter out, String msg) throws IOException {
     }
 
     /**
+     * This extension provides a Ping protocol that allows people to verify that the TcpSlaveAgentListener is alive.
+     * We also use this
```
Comment (daniel-beck) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: Comment seems incomplete. Also, while you're at it, `@since TODO`.

**Comment by stephenc** [comment]
Merging!

**Pull request merged**